packages/react-strict-dom/src/native/modules/useStyleTransition.js

```diff
@@ -92,319 +203,310 @@ export function useStyleTransition(style: ReactNativeStyle): ReactNativeStyle {
+      shouldUseNativeDriver: canUseNativeDriver(transitionStyle)
+    };
+  }, [
+    transitionDelay,
+    transitionDuration,
+    transitionStyle,
+    transitionTimingFunction
+  ]);
+
+  // effect to trigger a transition
+  // REMEMBER: it is super important that this effect's dependency array **only** contains the animated value
+  useEffect(() => {
+    if (animatedValue !== undefined) {
+      const { delay, duration, timingFunction, shouldUseNativeDriver } =
+        transitionMetadataRef.current;
+
       const animation = Animated.sequence([
-        Animated.delay(transitionDelay ?? 0),
-        Animated.timing(animatedRef.current, {
+        Animated.delay(delay),
+        Animated.timing(animatedValue, {
           toValue: 1,
-          duration: transitionDuration ?? 16,
-          easing: getEasingFunction(transitionTimingFunction),
-          useNativeDriver: canUseNativeDriver(transitionProperties)
+          duration,
+          easing: getEasingFunction(timingFunction),
+          useNativeDriver: shouldUseNativeDriver
         })
       ]);
-      animation.start(() => {
-        previousStyleRef.current = style;
-        animatedRef.current = new Animated.Value(0);
-      });
+      animation.start();
+
       return () => {
         animation.stop();
       };
     }
-  }, [
-    style,
-    transitionDelay,
-    transitionDuration,
-    transitionProperties,
-    transitionTimingFunction
-  ]);
+  }, [animatedValue]);
+
+  if (transitionStyleHasChanged(transitionStyle, currentStyle)) {
+    setCurrentStyle(style);
+    setPreviousStyle(currentStyle);
+    setAnimatedValue(new Animated.Value(0));
+    // This commit will be thrown away due to the above state setters so we can bail out early
+    return style;
+  }
 
-  if (transitionProperties == null) {
+  if (transitionStyle === undefined) {
     return style;
   }
 
-  const transitionStyle: AnimatedStyle = transitionProperties.reduce(
-    (animatedStyle, entry) => {
-      const { property, value } = entry;
+  const outputAnimatedStyle: AnimatedStyle = Object.entries(
+    transitionStyle
+  ).reduce<AnimatedStyle>((animatedStyle, [property, value]) => {
+    const startValue = previousStyle?.[property] ?? value;
 
-      const startValue = previousStyleRef.current?.[property];
+    if (animatedValue === undefined || startValue === value) {
+      animatedStyle[property] = value;
+    } else if (typeof value === 'number') {
+      animatedStyle[property] = animatedValue.interpolate({
+        inputRange: INPUT_RANGE,
+        outputRange: [+startValue, value]
+      });
+      return animatedStyle;
+    } else if (typeof value === 'string') {
+      animatedStyle[property] = animatedValue.interpolate({
+        inputRange: INPUT_RANGE,
+        outputRange: [String(startValue), value]
+      });
+      return animatedStyle;
+    } else if (property === 'transform' && Array.isArray(value)) {
+      const transforms = value;
+      const refTransforms = startValue;
```

Comment:
we can rename `refTransforms`/`startValue` to something like `prevTransforms`/`prevValue` now that it's state